docs/Dialects/Interop/RationaleInterop.md

```diff
@@ -0,0 +1,590 @@
+  %0 = systemc.cpp.new() : () -> !emitc.ptr<!emitc.opaque<"VBar">>
+  interop.return %0 : !emitc.ptr<!emitc.opaque<"VBar">>
+}
+```
+
+The `update` operation is similar to the `alloc` operation in that it has an
+interop mechanism attribute for the same reason and takes the state values as
+operands, but also passes them on to the body via block arguments using
+pass-by-value semantics. In addition to the state values, it also takes a variadic
+list of inputs and also passes them on to the body. The `return` inside the body
+then returns the result values after doing some computation inside the body.
+If the state needs to be mutated, it has to be a pointer type. If we need to be
+able to change the actual state value in the future, we could return updated
+states via the `return` operation (currently not allowed).
+
+```mlir
+%1 = interop.procedural.update cpp [%state](%x, %x) :
+  [!emitc.ptr<!emitc.opaque<"VBar">>](i32, i32) -> i32 {
+^bb0(%arg0: !emitc.ptr<!emitc.opaque<"VBar">>, %arg1: i32, %arg2: i32):
+  interop.return %5 : i32
+}
+```
+
+The `dealloc` operation shall be executed right before the state requested by
+the `alloc` operation is released. This allows the instance to do some cleanup,
+e.g., when the state type was a pointer and the instance performed some
+`malloc`. Structurally the operation is the same as the `update` operation, but
+without input and output values. The state is also passed by value.
+
+```mlir
+interop.procedural.dealloc cpp %state : !emitc.ptr<!emitc.opaque<"VBar">> {
+^bb0(%arg0: !emitc.ptr<!emitc.opaque<"VBar">>):
+  systemc.cpp.delete %arg0 : !emitc.ptr<!emitc.opaque<"VBar">>
+  interop.return
+}
+```
+
+### Interop Mechanisms
+
+A variety of interop mechanisms can be supported. This allows to perform interop
+on different abstraction levels and only fall back to C-foreign-functions when
+nothing else is supported.
+
+Currently supported interop mechanisms:
+* CFFI: C-foreign-functions, i.e., all interop operations are lowered to C
+  functions and function calls.
+* CPP: textual C++, i.e., the interop operations are lowered to the `systemc`
+  and `emitc` dialects and printed as C++ code. In the future, more dialects
+  (such as `scf`) could be supported.
+
+Adding a new interop mechanism requires changes to the `interop` dialect. At
+a minimum, the table-gen enum as to be modified and a bridging pattern has to
+be added.
+
+### Instance-side Lowering
+
+The instance-side lowering always has to happen before the container-side
+lowering since that pass should create the interop operations that will then
+be picked up by the container-side lowering pass and properly embedded in the
+context of the container operation.
+
+To illustrate how this works, consider a design represented in `hw`, `comb`, and
+`seq`. We want to simulate that design using Verilator and provide a SystemC
+wrapper (basically what Verilator itself can also do). As a first step, the
```

Comment:
It would be useful to highlight the Verilator option that this relates to, e.g., "SystemC Output Mode (`--sc`)".